lib/ReadsUtils/ReadsUtilsImpl.py

```diff
@@ -637,11 +698,359 @@ def validateFASTA(self, ctx, params):
+        with gzip.open(copy_file_path + '.gz', 'rb') as in_file:
+            with open(copy_file_path, 'w') as f:
+                f.write(in_file.read())
+        self.log('Unzipped file: %s' % copy_file_path + '.gz')
+
+    def _check_ftp_connection(self, user_name, password, domain, file_path, file_name):
+        """
+        _check_ftp_connection: ftp connection checker
+
+        params:
+        user_name: FTP user name
+        password: FTP user password
+        domain: FTP domain
+        file_path: target file directory
+        file_name: target file name 
+
+        """
+
+        try:
+            ftp = ftplib.FTP(domain)
+        except ftplib.all_errors, error:
+            raise ValueError("Cannot connect: %s" % error)
+        else:
+            try:
+                ftp.login(user_name, password)
+            except ftplib.all_errors, error:
+                raise ValueError("Cannot login: %s" % error)
+            else:
+                ftp.cwd(file_path)
+                if file_name in ftp.nlst():
+                    pass
+                else:
+                    raise ValueError("File %s does NOT exist in FTP path: %s" % (
+                        file_name, domain + '/' + file_path))
+
+    def _download_google_drive_link(self, file_url, copy_file_path):
+        """
+        _download_google_drive_link: Google Drive download link handler
+                                     file needs to be shared publicly 
+
+        params:
+        file_url: Google Drive download link
+        copy_file_path: output file saving path
+
+        """
+        # translate Google Drive URL for direct download
+        force_download_link_prefix = 'https://drive.google.com/uc?export=download&id='
+        file_id = file_url.partition('/d/')[-1].partition('/')[0]
+        force_download_link = force_download_link_prefix + file_id
+
+        self.log('Generating Google Drive direct download link\n from: %s\n to: %s' % (
+            file_url, force_download_link))
+        self._download_direct_download_link(
+            force_download_link, copy_file_path)
+
+    def _process_download(self, fwd, rev, reads_source, download_type, user_id):
+        """
+        _process_download: processing different type of downloads
+
+        fwd: forward shock_id if reads_source is 'shock'
+               forward url if reads_source is 'web'
+               forward file name in staging if reads_source is 'staging'
+        rev: reverse shock_id if reads_source is 'shock'
+               reverse url if reads_source is 'web'
+               reverse file name in staging if reads_source is 'staging'
+        reads_source: one of 'shock', 'web' or 'staging'
+        download_type: one of ['Direct Download', 'FTP', 'DropBox', 'Google Drive']
+        user_id: current token user
+        """
+
+        # return values
+        fwdname = None
+        revname = None
+        fwdpath = None
+        revpath = None
```

Comment:
👍 
remove fwdpath. 
revpath is not set when read_source == 'shock' and not revid.